### Données de 18/02/2019 00:00 jusqu'à 18/02/2019 16:22

In [107]:
#Charger les biblio
import numpy as np
import pandas as pd
from pandas import read_csv
import matplotlib.pyplot as plt
from matplotlib import cm
import seaborn as sns
import numpy as np
import string 
from subprocess import check_output
import seaborn as sns
from pylab import *
%matplotlib inline
sns.set()

In [162]:
# Lire la table avec 'header'
df = pd.read_csv('Hmvl2.csv', sep=';', engine='python')

In [163]:
# Remplacer les NaN par des zeros dans trame_hmvl
df.trame_hmvl.fillna('0', inplace=True)

In [164]:
#Identifier les donnees de chaque stations
groupe = df.groupby('code')
#Table de donnees de la station M7i
df = groupe.get_group('M7i')

In [165]:
# Reindexer la dataframe
df.index = range(len(df))

In [166]:
df.head(3)

code    id  equipement_id           date_reception   trame_hmvl
0  M7i  6460           6460  2019-02-18 00:00:03.049            0
1  M7i  6460           6460  2019-02-18 00:00:15.049  14148093033
2  M7i  6460           6460  2019-02-18 00:00:21.049            0

Séparons d'abords la colonne 'trame_hmvl' par cases de 11 caractères.

In [167]:
#changeaons les noms de colonnes
df.columns = ['code', 'id','equipement_id', 'date','trame']

Déterminer le nombre maximale de caractère :

In [168]:
# On ajoute une colonne Count pour compter chacune
df['Count'] = df['trame'].str.len()

C:\Users\oumayma.doughmi\ANACONDA\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [169]:
df.Count.max()

154

In [170]:
# Supprimons la colonne
df = df.drop(['Count'], axis=1)

 154 est le nombre maximale de caractère, donc il faut créer 14 nouveaux champs qui contient les 11 caractères pour chaque véhicules.

réalisons une boucle qui séparer toutes les colonnes à la fois

In [171]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10','x11','x12', 'x13']

In [172]:
# Créer les tables 
for i in range(len(liste)):
    df[liste[i]] = df["trame"].str.slice(11*i,1*(i+11))

In [173]:
df.head()

code    id  equipement_id                     date        trame  \
0  M7i  6460           6460  2019-02-18 00:00:03.049            0   
1  M7i  6460           6460  2019-02-18 00:00:15.049  14148093033   
2  M7i  6460           6460  2019-02-18 00:00:21.049            0   
3  M7i  6460           6460  2019-02-18 00:00:26.549  05354096031   
4  M7i  6460           6460  2019-02-18 00:00:33.049            0   

            x0 x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13  
0            0                                             
1  14148093033                                             
2            0                                             
3  05354096031                                             
4            0

Dans le but d'avoir une ligne par véhicules, nous allons créer des tables $x_{i}$ en supprimant les $x_{j}$ avec $j \ne i$.

In [175]:
# Supprimer les colonnes xj
df0 = df.drop(['x3','x1','x2','x4','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
# df1
df1 = df.drop(['x3','x0','x2','x4','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)

In [174]:
l = ['voie', 'horodatage', 'vitesse', 'longueur']

In [ ]:
# Créer les tables 
for i in range(len(liste)):
    df0[liste[i]] = df["x0"].str.slice(1*i,1*(i+1))

In [160]:
# Premier champs x0
df0["voie"] = df0["x0"].str.slice(0,1)
df0["horodatage"] = df0["x0"].str.slice(1,5)
df0["vitesse"] = df0["x0"].str.slice(5,8)
df0["longueur"] = df0["x0"].str.slice(8,11)

In [161]:
df0.head()

code    id  equipement_id                     date           x0 voie  \
0  M7i  6460           6460  2019-02-18 00:00:03.049            0    0   
1  M7i  6460           6460  2019-02-18 00:00:15.049  14148093033    1   
2  M7i  6460           6460  2019-02-18 00:00:21.049            0    0   
3  M7i  6460           6460  2019-02-18 00:00:26.549  05354096031    0   
4  M7i  6460           6460  2019-02-18 00:00:33.049            0    0   

  horodatage vitesse longueur  
0                              
1       4148     093      033  
2                              
3       5354     096      031  
4

In [142]:
# Supprimer les colonnes xj
df4 = df.drop(['x0','x1','x2','x3','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)

In [177]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10','x11','x12', 'x13']
liste2 = ["df0","df1","df2","df3","df4","df5","df6","df7","df8","df9",'df10','df11','df12', 'df13']

for i, f in zip(liste, liste2):
    f["voie"] = f[i].str.slice(0,1)
    f["horodatage"] = f[i].str.slice(1,5)
    f["vitesse"] = f[i].str.slice(5,8)
    f["longueur"] = f[i].str.slice(8,11)
    

TypeError: string indices must be integers

In [143]:
# faisons la même chose pour les autres champs
# df1
df1["voie"] = df1["x1"].str.slice(0,1)
df1["horodatage"] = df1["x1"].str.slice(1,5)
df1["vitesse"] = df1["x1"].str.slice(5,8)
df1["longueur"] = df1["x1"].str.slice(8,11)
# df2
# Supprimer les colonnes xj
df2 = df.drop(['x3','x1','x0','x4','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
df2["voie"] = df2["x2"].str.slice(0,1)
df2["horodatage"] = df2["x2"].str.slice(1,5)
df2["vitesse"] = df2["x2"].str.slice(5,8)
df2["longueur"] = df2["x2"].str.slice(8,11)
# df3
# Supprimer les colonnes xj
df3 = df.drop(['x0','x1','x2','x4','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
df3["voie"] = df3["x3"].str.slice(0,1)
df3["horodatage"] = df3["x3"].str.slice(1,5)
df3["vitesse"] = df3["x3"].str.slice(5,8)
df3["longueur"] = df3["x3"].str.slice(8,11)
# df4
df4["voie"] = df4["x4"].str.slice(0,1)
df4["horodatage"] = df4["x4"].str.slice(1,5)
df4["vitesse"] = df4["x4"].str.slice(5,8)
df4["longueur"] = df4["x4"].str.slice(8,11)

In [144]:

# Supprimer les colonnes xj
df5 = df.drop(['x0','x1','x2','x4','x3','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
# Supprimer les colonnes xj
df6 = df.drop(['x0','x1','x2','x4','x5','x3','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
# Supprimer les colonnes xj
df7 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
# Supprimer les colonnes xj
df8 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)

In [145]:
# df5
df5["voie"] = df5["x5"].str.slice(0,1)
df5["horodatage"] = df5["x5"].str.slice(1,5)
df5["vitesse"] = df5["x5"].str.slice(5,8)
df5["longueur"] = df5["x5"].str.slice(8,11)
# df6
df6["voie"] = df6["x6"].str.slice(0,1)
df6["horodatage"] = df6["x6"].str.slice(1,5)
df6["vitesse"] = df6["x6"].str.slice(5,8)
df6["longueur"] = df6["x6"].str.slice(8,11)
# df7
df7["voie"] = df7["x7"].str.slice(0,1)
df7["horodatage"] = df7["x7"].str.slice(1,5)
df7["vitesse"] = df7["x7"].str.slice(5,8)
df7["longueur"] = df7["x7"].str.slice(8,11)
# df8
df8["voie"] = df8["x8"].str.slice(0,1)
df8["horodatage"] = df8["x8"].str.slice(1,5)
df8["vitesse"] = df8["x8"].str.slice(5,8)
df8["longueur"] = df8["x8"].str.slice(8,11)

In [146]:
df9 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x8', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)

df10 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x9', 'x8', 'x11','x12', 'x13', 'trame'], axis=1)

df11 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x9', 'x10', 'x8','x12', 'x13', 'trame'], axis=1)

df12 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x8', 'x10', 'x11','x9', 'x13', 'trame'], axis=1)

df13 = df.drop(['x0','x1','x2','x4','x5','x6','x3',
               'x7', 'x9', 'x8', 'x11','x12', 'x10', 'trame'], axis=1)

In [147]:
# df
df9["voie"] = df9["x9"].str.slice(0,1)
df9["horodatage"] = df9["x9"].str.slice(1,5)
df9["vitesse"] = df9["x9"].str.slice(5,8)
df9["longueur"] = df9["x9"].str.slice(8,11)
# df10
df10["voie"] = df10["x10"].str.slice(0,1)
df10["horodatage"] = df10["x10"].str.slice(1,5)
df10["vitesse"] = df10["x10"].str.slice(5,8)
df10["longueur"] = df10["x10"].str.slice(8,11)
# df12
df12["voie"] = df12["x12"].str.slice(0,1)
df12["horodatage"] = df12["x12"].str.slice(1,5)
df12["vitesse"] = df12["x12"].str.slice(5,8)
df12["longueur"] = df12["x12"].str.slice(8,11)
# df11
df11["voie"] = df["x11"].str.slice(0,1)
df11["horodatage"] = df["x11"].str.slice(1,5)
df11["vitesse"] = df["x11"].str.slice(5,8)
df11["longueur"] = df["x11"].str.slice(8,11)
# df13
df13["voie"] = df13["x13"].str.slice(0,1)
df13["horodatage"] = df13["x13"].str.slice(1,5)
df13["vitesse"] = df13["x13"].str.slice(5,8)
df13["longueur"] = df13["x13"].str.slice(8,11)

In [79]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10','x11','x12', 'x13']
#,"x8","x9",'x10','x11','x12', 'x13'


In [80]:
# Créer les tables 
for i in range(len(liste)):
    # Supprimer les colonnes différentes de :
    liste[i] = df.drop(df.columns.difference([liste[i],'code', 'id', 'equipement_id', 'date']), 1, inplace=True)

In [81]:
x1

NameError: name 'x1' is not defined

In [70]:
for i in liste :
    # Supprimer les colonnes différentes de :
    i = df.drop(df.columns.difference([i,'code', 'id', 'equipement_id', 'date']), 1, inplace=True)
    i["voie"] = i[i].str.slice(0,1)
    i["horodatage"] = i[i].str.slice(1,5)
    i["vitesse"] = i[i].str.slice(5,8)
    i["longueur"] = i[i].str.slice(8,11)

TypeError: 'NoneType' object is not subscriptable

In [ ]:
# df2
# Supprimer les colonnes xj
x0 = df.drop(['x3','x1','x0','x4','x5','x6','x7',
               'x8', 'x9', 'x10', 'x11','x12', 'x13', 'trame'], axis=1)
df2["voie"] = df2["x2"].str.slice(0,1)
df2["horodatage"] = df2["x2"].str.slice(1,5)
df2["vitesse"] = df2["x2"].str.slice(5,8)
df2["longueur"] = df2["x2"].str.slice(8,11)

In [154]:
 df0.rename(columns={'x0':'x'}, inplace=True)
df1.rename(columns={'x1':'x'}, inplace=True)
df2.rename(columns={'x2':'x'}, inplace=True)
df3.rename(columns={ 'x3':'x'}, inplace=True)
df4.rename(columns={ 'x4':'x'}, inplace=True)
df5.rename(columns={ 'x5':'x'}, inplace=True)
df6.rename(columns={'x6':'x'}, inplace=True)
df7.rename(columns={ 'x7':'x'}, inplace=True)
df8.rename(columns={ 'x8':'x'}, inplace=True)
df9.rename(columns={ 'x9':'x'}, inplace=True)
df10.rename(columns={ 'x10':'x'}, inplace=True)
df11.rename(columns={ 'x11':'x'}, inplace=True)
df12.rename(columns={ 'x12':'x'}, inplace=True)
df13.rename(columns={'x13':'x'}, inplace=True)

In [158]:
df1

code    id  equipement_id                     date  x voie horodatage  \
0     M7i  6460           6460  2019-02-18 00:00:03.049                      
1     M7i  6460           6460  2019-02-18 00:00:15.049                      
2     M7i  6460           6460  2019-02-18 00:00:21.049                      
3     M7i  6460           6460  2019-02-18 00:00:26.549                      
4     M7i  6460           6460  2019-02-18 00:00:33.049                      
5     M7i  6460           6460  2019-02-18 00:00:38.549                      
6     M7i  6460           6460  2019-02-18 00:00:44.631                      
7     M7i  6460           6460  2019-02-18 00:00:50.548                      
8     M7i  6460           6460  2019-02-18 00:00:56.549                      
9     M7i  6460           6460  2019-02-18 00:01:03.049                      
10    M7i  6460           6460  2019-02-18 00:01:09.049                      
11    M7i  6460           6460   2019-02-18 00:01:15.05                      
12    M7i  6460           6460  2019-02-18 00:01:21.049  1    1              
13    M7i  6460           6460  2019-02-18 00:01:27.049                      
14    M7i  6460           6460  2019-02-18 00:01:33.049                      
15    M7i  6460           6460  2019-02-18 00:01:39.049                      
16    M7i  6460           6460   2019-02-18 00:01:51.05                      
17    M7i  6460           6460  2019-02-18 00:01:57.049                      
18    M7i  6460           6460  2019-02-18 00:02:08.549                      
19    M7i  6460           6460  2019-02-18 00:02:15.049                      
20    M7i  6460           6460  2019-02-18 00:02:21.048                      
21    M7i  6460           6460  2019-02-18 00:02:32.549  0    0              
22    M7i  6460           6460  2019-02-18 00:02:39.049                      
23    M7i  6460           6460  2019-02-18 00:02:57.049                      
24    M7i  6460           6460  2019-02-18 00:03:03.049                      
25    M7i  6460           6460  2019-02-18 00:03:09.049                      
26    M7i  6460           6460  2019-02-18 00:03:15.049                      
27    M7i  6460           6460  2019-02-18 00:03:21.049                      
28    M7i  6460           6460  2019-02-18 00:03:27.049                      
29    M7i  6460           6460  2019-02-18 00:03:33.048                      
...   ...   ...            ...                      ... ..  ...        ...   
7124  M7i  6460           6460  2019-02-18 16:19:52.049  0    0              
7125  M7i  6460           6460  2019-02-18 16:19:58.049  2    2              
7126  M7i  6460           6460  2019-02-18 16:20:04.049  1    1              
7127  M7i  6460           6460  2019-02-18 16:20:10.049  0    0              
7128  M7i  6460           6460  2019-02-18 16:20:17.085  0    0              
7129  M7i  6460           6460  2019-02-18 16:20:22.049  2    2              
7130  M7i  6460           6460  2019-02-18 16:20:28.085  0    0              
7131  M7i  6460           6460  2019-02-18 16:20:34.049  1    1              
7132  M7i  6460           6460  2019-02-18 16:20:40.049  2    2              
7133  M7i  6460           6460  2019-02-18 16:20:46.049  0    0              
7134  M7i  6460           6460  2019-02-18 16:20:52.049  0    0              
7135  M7i  6460           6460  2019-02-18 16:20:58.565  0    0              
7136  M7i  6460           6460  2019-02-18 16:21:04.049  1    1              
7137  M7i  6460           6460  2019-02-18 16:21:10.549  2    2              
7138  M7i  6460           6460  2019-02-18 16:21:16.049  1    1              
7139  M7i  6460           6460  2019-02-18 16:21:22.049  2    2              
7140  M7i  6460           6460  2019-02-18 16:21:28.049  2    2              
7141  M7i  6460           6460  2019-02-18 16:21:34.049  0    0              
7142  M7i  6460           6460  2019-02-18 16:21:40.049  2    2              
7143  M7i  6460           6

In [156]:
frames = [df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12, df13]

result = pd.concat(frames)

In [157]:
result

code    id  equipement_id                     date  x voie horodatage  \
0     M7i  6460           6460  2019-02-18 00:00:03.049                      
1     M7i  6460           6460  2019-02-18 00:00:15.049                      
2     M7i  6460           6460  2019-02-18 00:00:21.049                      
3     M7i  6460           6460  2019-02-18 00:00:26.549                      
4     M7i  6460           6460  2019-02-18 00:00:33.049                      
5     M7i  6460           6460  2019-02-18 00:00:38.549                      
6     M7i  6460           6460  2019-02-18 00:00:44.631                      
7     M7i  6460           6460  2019-02-18 00:00:50.548                      
8     M7i  6460           6460  2019-02-18 00:00:56.549                      
9     M7i  6460           6460  2019-02-18 00:01:03.049                      
10    M7i  6460           6460  2019-02-18 00:01:09.049                      
11    M7i  6460           6460   2019-02-18 00:01:15.05                      
12    M7i  6460           6460  2019-02-18 00:01:21.049  1    1              
13    M7i  6460           6460  2019-02-18 00:01:27.049                      
14    M7i  6460           6460  2019-02-18 00:01:33.049                      
15    M7i  6460           6460  2019-02-18 00:01:39.049                      
16    M7i  6460           6460   2019-02-18 00:01:51.05                      
17    M7i  6460           6460  2019-02-18 00:01:57.049                      
18    M7i  6460           6460  2019-02-18 00:02:08.549                      
19    M7i  6460           6460  2019-02-18 00:02:15.049                      
20    M7i  6460           6460  2019-02-18 00:02:21.048                      
21    M7i  6460           6460  2019-02-18 00:02:32.549  0    0              
22    M7i  6460           6460  2019-02-18 00:02:39.049                      
23    M7i  6460           6460  2019-02-18 00:02:57.049                      
24    M7i  6460           6460  2019-02-18 00:03:03.049                      
25    M7i  6460           6460  2019-02-18 00:03:09.049                      
26    M7i  6460           6460  2019-02-18 00:03:15.049                      
27    M7i  6460           6460  2019-02-18 00:03:21.049                      
28    M7i  6460           6460  2019-02-18 00:03:27.049                      
29    M7i  6460           6460  2019-02-18 00:03:33.048                      
...   ...   ...            ...                      ... ..  ...        ...   
7124  M7i  6460           6460  2019-02-18 16:19:52.049                      
7125  M7i  6460           6460  2019-02-18 16:19:58.049                      
7126  M7i  6460           6460  2019-02-18 16:20:04.049                      
7127  M7i  6460           6460  2019-02-18 16:20:10.049                      
7128  M7i  6460           6460  2019-02-18 16:20:17.085                      
7129  M7i  6460           6460  2019-02-18 16:20:22.049                      
7130  M7i  6460           6460  2019-02-18 16:20:28.085                      
7131  M7i  6460           6460  2019-02-18 16:20:34.049                      
7132  M7i  6460           6460  2019-02-18 16:20:40.049                      
7133  M7i  6460           6460  2019-02-18 16:20:46.049                      
7134  M7i  6460           6460  2019-02-18 16:20:52.049                      
7135  M7i  6460           6460  2019-02-18 16:20:58.565                      
7136  M7i  6460           6460  2019-02-18 16:21:04.049                      
7137  M7i  6460           6460  2019-02-18 16:21:10.549                      
7138  M7i  6460           6460  2019-02-18 16:21:16.049                      
7139  M7i  6460           6460  2019-02-18 16:21:22.049                      
7140  M7i  6460           6460  2019-02-18 16:21:28.049                      
7141  M7i  6460           6460  2019-02-18 16:21:34.049                      
7142  M7i  6460           6460  2019-02-18 16:21:40.049                      
7143  M7i  6460           6

In [135]:
type(df1)

NoneType

In [136]:
df13.rename(columns={'x': 'x13'}, inplace=True)

AttributeError: 'NoneType' object has no attribute 'rename'